l3_footstep_planning_plugins/src/base/hlut_heuristic_plugin.cpp

```diff
@@ -0,0 +1,244 @@
+    // if there is already a lower heuristic value for this index, skip it
+    if (hlut_.getHeuristicEntry(current_index) <= min_element.heuristic_distance)
+      continue;
+
+    // insert the new heuristic value into the HLUT
+    hlut_.setHeuristicEntry(current_index, min_element.heuristic_distance);
+
+    // stop if the start index has been reached
+    if (current_index == start_index)
+      break;
+
+    std::vector<l3::PositionIndex> valid_neighbors = getValidNeighbors(getNeighbors(current_index));
+
+    for (auto& valid_neighbor : valid_neighbors)
+    {
+      // insert the neighbor with its heuristic value into the min heap
+      considered_next.push(computeHLUTEntryOfNeighbor(valid_neighbor, min_element));
+    }
+  }
+
+  if (visualize_)
+    visualizeHLUT();
+}
+
+double HLUTHeuristicPlugin::getHeuristicValue(const State& from, const State& to, const State& start, const State& goal) const
+{
+  // get the floating bases
+  FloatingBase::ConstPtr from_fb = from.getFloatingBase(l3::BaseInfo::MAIN_BODY_IDX);
+  FloatingBase::ConstPtr to_fb = to.getFloatingBase(l3::BaseInfo::MAIN_BODY_IDX);
+
+  return getHeuristicValue(*from_fb, *to_fb, start, goal);
+}
+
+double HLUTHeuristicPlugin::getHeuristicValue(const Foothold& from, const Foothold& /*to*/, const State& /*start*/, const State& /*goal*/) const
+{
+  // get the heuristic value from the HLUT for the given position
+  return hlut_.getHeuristicEntry(l3::Position2D(from.x(), from.y()));
+}
+
+double HLUTHeuristicPlugin::getHeuristicValue(const FloatingBase& from, const FloatingBase& /*to*/, const State& /*start*/, const State& /*goal*/) const
+{
+  // get the heuristic value from the HLUT for the given position
+  return hlut_.getHeuristicEntry(l3::Position2D(from.x(), from.y()));
+}
+
+HeuristicLookupTable HLUTHeuristicPlugin::initializeHLUT(const FloatingBase& start_fb, const FloatingBase& goal_fb, const l3::Position2D& goal_pos) const
+{
+  int size_x = size_x_;
+  int size_y = size_y_;
+
+  // check if start lies within the HLUT with the specified size.
+  // if not, resize the HLUT
+  if (euclideanDistance(start_fb.x(), 0.0, goal_fb.x(), 0.0) > size_x_ * 0.5)
+  {
+    size_x = 2 * static_cast<int>(std::ceil(euclideanDistance(start_fb.x(), 0.0, goal_fb.x(), 0.0))) + 1;
+    ROS_WARN("[%s]: HLUT size_x is too small. Increasing to %d.", getName().c_str(), size_x);
+  }
+  if (euclideanDistance(0.0, start_fb.y(), 0.0, goal_fb.y()) > size_y_ * 0.5)
+  {
+    size_y = 2 * static_cast<int>(std::ceil(euclideanDistance(0.0, start_fb.y(), 0.0, goal_fb.y()))) + 1;
+    ROS_WARN("[%s]: HLUT size_y is too small. Increasing to %d.", getName().c_str(), size_y);
+  }
+
+  // initialize HLUT
+  return HeuristicLookupTable(resolution_, Eigen::Vector2i(size_x, size_y), goal_pos);
+}
+
+std::vector<l3::PositionIndex> HLUTHeuristicPlugin::getValidNeighbors(const std::vector<l3::PositionIndex>& neighbors) const
+{
+  std::vector<l3::PositionIndex> valid_neighbors = {};
```

Comment:
You can reserve (not resize!) the memory block for the vector as you already know the upper limit of neighbors. This may save expensive automatic memory reallocation.